src/rasta/c/rasta_new.c

```diff
@@ -1128,408 +1152,266 @@ void handle_retrdata(struct rasta_receive_handle *h, struct rasta_connection *co
 /*
  * threads
  */
-void * receive_thread(void * handle){
-    // enable the possibility to cancel this thread
-    pthread_setcancelstate(PTHREAD_CANCEL_ENABLE, NULL);
-
-    // set type to async mode, i.e the thread will be cancelled as soon as pthread_cancel is called
-    pthread_setcanceltype(PTHREAD_CANCEL_DEFERRED, NULL);
-
-    struct rasta_receive_handle *h = (struct rasta_receive_handle*)handle;
-    logger_log(h->logger, LOG_LEVEL_DEBUG, "RaSTA RECEIVE", "Receive thread started");
-
-    // run as long as the running flag is set to true
-    while (*h->running){
-        logger_log(h->logger, LOG_LEVEL_DEBUG, "RaSTA RECEIVE", "Waiting for incoming packets...");
-
-        // wait for incoming packets
-        struct RastaPacket receivedPacket = redundancy_mux_retrieve_all(h->mux);
-
-        logger_log(h->logger, LOG_LEVEL_DEBUG, "RaSTA RECEIVE", "Received packet %d from %d to %d", receivedPacket.type, receivedPacket.sender_id, receivedPacket.receiver_id);
-
 
+char on_readable_event(void * handle) {
+    struct rasta_receive_handle *h = (struct rasta_receive_handle*) handle;
 
-        struct rasta_connection * con = rastalist_getConnectionByRemote(h->connections, receivedPacket.sender_id);
-
-        //new client request
-        if (receivedPacket.type == RASTA_TYPE_CONNREQ){
-            if (con != 0) {
-                pthread_mutex_lock(&con->lock);
-                handle_conreq(h,con,receivedPacket);
-                pthread_mutex_unlock(&con->lock);
-
-            } else {
-                handle_conreq(h,0,receivedPacket);
-            }
-
-            freeRastaByteArray(&receivedPacket.data);
-            continue;
-        }
-
-
-        if (con == 0) {
-            logger_log(h->logger, LOG_LEVEL_DEBUG, "RaSTA RECEIVE", "Received packet (%d) from unknown source %d", receivedPacket.type, receivedPacket.sender_id);
-            //received packet from unknown source
-            //TODO: can these packets be ignored?
+    // wait for incoming packets
+    struct RastaPacket receivedPacket;
+    if (!redundancy_mux_try_retrieve_all(h->mux, &receivedPacket)) {
+        return 0;
+    }
 
-            freeRastaByteArray(&receivedPacket.data);
-            continue;
-        }
+    logger_log(h->logger, LOG_LEVEL_DEBUG, "RaSTA RECEIVE", "Received packet %d from %d to %d", receivedPacket.type, receivedPacket.sender_id, receivedPacket.receiver_id);
 
-        //lock this connection
-        pthread_mutex_lock(&con->lock);
+    int con_id = rastalist_getConnectionId(h->connections, receivedPacket.sender_id);
+    struct rasta_connection* con;
+    //new client request
+    if (receivedPacket.type == RASTA_TYPE_CONNREQ){
```

Comment:
formatting: possibly add spaces after `//` and before curly brace (i.e. `if () {`)